# Random Forest Classifier 
- It uses training data to create multiple data set using "BootStraping Method"
- majority vote is final predicted class.
- StratifiedKFold select and ensures balanced class selction it will select with equal representation
- random froest operates in the principle of ensembble learning where multiple weak models or avg 

In [3]:
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import KFold, StratifiedKFold

In [5]:
dataframe = pd.read_csv("diabetes.csv")
dataframe

preg  plas  pres  skin  test  mass   pedi  age  class
0       6   148    72    35     0  33.6  0.627   50      1
1       1    85    66    29     0  26.6  0.351   31      0
2       8   183    64     0     0  23.3  0.672   32      1
3       1    89    66    23    94  28.1  0.167   21      0
4       0   137    40    35   168  43.1  2.288   33      1
..    ...   ...   ...   ...   ...   ...    ...  ...    ...
763    10   101    76    48   180  32.9  0.171   63      0
764     2   122    70    27     0  36.8  0.340   27      0
765     5   121    72    23   112  26.2  0.245   30      0
766     1   126    60     0     0  30.1  0.349   47      1
767     1    93    70    31     0  30.4  0.315   23      0

[768 rows x 9 columns]

In [9]:
dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   preg    768 non-null    int64  
 1   plas    768 non-null    int64  
 2   pres    768 non-null    int64  
 3   skin    768 non-null    int64  
 4   test    768 non-null    int64  
 5   mass    768 non-null    float64
 6   pedi    768 non-null    float64
 7   age     768 non-null    int64  
 8   class   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


#### **Overview of Pima Indian diabetes dataset**

-Preg	Number of times pregnant	Numeric	[0, 17]

-Gluc	Plasma glucose concentration at 2 Hours in an oral glucose tolerance test (GTIT)	Numeric	[0, 199]

-BP	Diastolic Blood Pressure (mm Hg)	Numeric	[0, 122]

-Skin	Triceps skin fold thickness (mm)	Numeric	[0, 99]

-Insulin	2-Hour Serum insulin (µh/ml)	Numeric	[0, 846]

-BMI	Body mass index [weight in kg/(Height in m)]	Numeric	[0, 67.1]

-DPF	Diabetes pedigree function	Numeric	[0.078, 2.42]

-Age	Age (years)	Numeric	[21, 81]

-Outcome	Binary value indicating non-diabetic /diabetic	Factor	[0,1]

In [28]:
from sklearn.model_selection import KFold, StratifiedKFold
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import RandomForestClassifier

X = dataframe.iloc[:,0:8]
Y = dataframe.iloc[:,8]

kfold = StratifiedKFold(n_splits=10,random_state= 3,shuffle = True)

model = RandomForestClassifier(n_estimators = 200,random_state = 20,max_depth = None)# here 200 decison tree opinion is taken 
results = cross_val_score(model, X, Y,cv=kfold)
print(results)
print(results.mean())

[0.75324675 0.76623377 0.75324675 0.80519481 0.87012987 0.84415584
 0.74025974 0.68831169 0.72368421 0.77631579]
0.7720779220779221


In [26]:
dataframe["class"].value_counts()

class
0    500
1    268
Name: count, dtype: int64

# Handling imbalanced classes
# Hyper parameter tuning in random Forest
### Tuning a random forest model involves adjusting several parameters to optimize its performance 
1.  **no.of trees(n_estimators):** this parameter specifies the no.of trees in the forest.Generaly    but also increases the computational cost and time
2.  **Max depth of trees(Max_depth):** we grow the decision tree until we get all the leaf nodes.
3. **Min Sample Split(min_samples_split):**
4.  **Min Samples Leaf(min_samples_leaf):**
5.  **Max Fetures(max_features):**
6.  **Bootstrap(bootstarp):**
By careful tuning these parameters based on cross-validation performance, one can significantly improve 

### (11/03/2025)
